### ISSUE TITLE
Long lived clients from grpc eventual oom an akka grpc/http2 server

### ISSUE THREAD

**Issue by Marcus-Rosti** [akka/akka-http#2551]
Please refer to https://github.com/akka/akka-grpc/issues/591

In brief, grpc encourages creating a long lived client to complete all http2 requests. However, these long lived clients don't allow GC to happen on the OutStreams from https://github.com/akka/akka-http/blob/master/akka-http2-support/src/main/scala/akka/http/impl/engine/http2/Http2Multiplexer.scala#158

Since the client is never closed, it keeps all of the objects active, and thus we eventually run out of heap space.

**Comment by jrudolph** [comment]
Thanks for the report, @Marcus-Rosti. It would be good if you could provide some reproducer, so we could debug the issue. Otherwise, DEBUG log output for the HTTP/2 parts would be helpful as well.

It seems, in the file you linked, `streamFor` can create (or recreate) OutStreams whenever requested, even if the stream has been closed before. It would be good to find out which frame would have recreated a stream that was closed before.

Some background information:

There's some inherent difficulty in implementing HTTP/2 that is to keep track of the state of streams, especially when they are inactive, i.e. still idle or already closed. In both cases, you cannot just throw away all state because you might want to keep prioritization information even for closed streams (currently, we don't implement much of prioritization, so we should not keep prioritization information for closed streams).

One important thing, we don't seem to implement yet, is to keep a running counter of what the highest stream id is that we have seen yet. Because of https://tools.ietf.org/html/rfc7540#section-5.1.1:

```
   The first use of a new stream identifier implicitly closes all
   streams in the "idle" state that might have been initiated by that
   peer with a lower-valued stream identifier.  For example, if a client
   sends a HEADERS frame on stream 7 without ever sending a frame on
   stream 5, then stream 5 transitions to the "closed" state when the
   first frame for stream 7 is sent or received.
```

we can then maintain that any stream that's not in the `outStreams` map is either idle or closed and distinguish those by comparing to the highest established stream id. We can then make sure not to put those streams back in the map if they have already been closed.

**Comment by Marcus-Rosti** [comment]
@jrudolph I have a repro case here and a screenshot of the effect. https://github.com/Marcus-Rosti/akka-grpc-sample-kubernetes-scala/pull/1

**Comment by Marcus-Rosti** [comment]
also that's very informative! Thank you.

I think the idea behind this is solid; however, it makes it impossible to release a service that will receive sustained traffic. If it's never able to release those objects the JVM will always oom eventually.

**Comment by Marcus-Rosti** [comment]
As per my example it happens also with unary traffic

**Comment by jrudolph** [comment]
Thanks, I verified your reproducer. The cause seems to be something different in your case. It seems the `incomingStreams` map keeps all entries in the state `HalfClosedRemote`. I.e. we are still waiting for the response to be finished. Maybe we are missing to report back from the outgoing side that the stream is complete?

**Comment by jrudolph** [comment]
@Marcus-Rosti I don't see any OutStreams remaining in the heap, so there might be another issue with those? If you observed those as well, it could make sense to add this case to the reproducer so I could have a look at this as well.

**Comment by Marcus-Rosti** [comment]
Right my case doesn't show the outstreams. If we change the call to the streaming case we'll see it. I can add that for you.

**Comment by raboof** [comment]
Looking at the code I indeed see the risk of incoming streams remaining in `HalfClosedRemote` state. I started on a fix in the `closeIncomingHttp2streams` branch (#2585) - it seems to work functionally speaking, but I haven't been able to reproduce the problem in VisualVM very convincingly. Could you give this branch a try?

**Comment by jrudolph** [comment]
We have fixed the incoming state part. 

Regarding

> One important thing, we don't seem to implement yet, is to keep a running counter of what the highest stream id is that we have seen yet. Because of tools.ietf.org/html/rfc7540#section-5.1.1:

We have implemented that logic already for the incoming side. However, it's unclear if it could still happen `outStreams`. I created https://github.com/akka/akka-http/issues/2587 to verify/fix that issue.

Now that we've merged #2585, @Marcus-Rosti can you check if that already fixes your issue or if you still see Outstreams (in which case we must also fix #2587)?

**Comment by jrudolph** [comment]
> Now that we've merged #2585, @Marcus-Rosti can you check if that already fixes your issue or if you still see Outstreams (in which case we must also fix #2587)?

We publish snapshots for every master commit at https://dl.bintray.com/akka/snapshots. The version for this fix will be 10.1.8+52-50c5faa6.
